feat: detect docker version and Playwright version mismatch

aslushnikov opened the pull request:
This patch prints a friendly instructions in case Docker image version
mismatches Playwright version and there are missing browser
dependencies.

With this patch, Playwright will yield the following error:

```
root@f0774d2b2097:~# node a.mjs
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

browserType.launch:
╔════════════════════════════════════════════════════════════════════════════════════════════╗
║ Host system is missing dependencies to run browsers.                                       ║
║ This is most likely due to docker image version not matching Playwright version:           ║
║ - Playwright: 1.22.0                                                                       ║
║ -     Docker: 1.21.0                                                                       ║
║                                                                                            ║
║ Either:                                                                                    ║
║ - (recommended) use docker image "mcr.microsoft.com/playwright:v1.22.0-focal"              ║
║ - (alternative 1) run the following command inside docker to install missing dependencies: ║
║                                                                                            ║
║     npx playwright install-deps                                                            ║
║                                                                                            ║
║ - (alternative 2) use Aptitude inside docker:                                              ║
║                                                                                            ║
║     apt-get install libgbm1                                                                ║
║                                                                                            ║
║ <3 Playwright Team                                                                         ║
╚════════════════════════════════════════════════════════════════════════════════════════════╝
    at file:///root/a.mjs:3:10 {
  name: 'Error'
}```

References #12796

mxschmitt commented on utils/docker/build.sh:
```diff
@@ -37,4 +37,8 @@ else
 fi
 
-docker build --platform "${PLATFORM}" -t "$3" -f "Dockerfile.$2" .
+PW_VERSION=$(node ../../utils/workspace.js --get-version)
```
we do the same also here, lets replace it also with --get-version? https://github.com/microsoft/playwright/blob/76c37f1977bfdb31165e09db61b4c43d95a9ac7c/utils/docker/publish_docker.sh#L10

dgozman commented on packages/playwright-core/src/utils/dependencies.ts:
```diff
@@ -179,6 +179,28 @@ export async function validateDependenciesLinux(sdkLanguage: string, linuxLddDir
+    throw new Error('\n' + utils.wrapInASCIIBox([
+      `Host system is missing a few dependencies to run browsers.`,
+      ``,
+      `This is due to docker image version not matching Playwright version:`,
```
"most likely"

dgozman commented on packages/playwright-core/src/utils/dependencies.ts:
```diff
@@ -179,6 +179,28 @@ export async function validateDependenciesLinux(sdkLanguage: string, linuxLddDir
+      `- (recommended) use docker image "${requiredDockerImage}"`,
+      `- (alternative) run the following command inside docker to install missing dependencies:`,
+      ``,
+      `    ${maybeSudo}${buildPlaywrightCLICommand(sdkLanguage, 'install-deps')}`,
```
I see "Unhappy path" below, so this command is not always the right one? Let's be careful.

dgozman reviewed (commented):
Nice!

mxschmitt commented:
Instead of making the docker build process more complicated, we could think about a hidden command in our CLI: "mark-pw-version-to-os" which will store the current version on the file system, e.g. `ms-playwright` folder. 

On the languages. like .NET or Java etc. its often quite complicated to get the version, you need to read XML. And in Python there is no version in the actual code.

This would enable us to just add `npx playwright mark-pw-version-to-os` in our docker images and we are golden!

aslushnikov committed:
feat: detect docker version and Playwright version mismatch

This patch prints a friendly instructions in case Docker image version
mismatches Playwright version and there are missing browser
dependencies.

With this patch, Playwright will yield the following error:

```
root@f0774d2b2097:~# node a.mjs
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

browserType.launch:
╔════════════════════════════════════════════════════════════════════════════════════════════╗
║ Host system is missing dependencies to run browsers.                                       ║
║ This is most likely due to docker image version not matching Playwright version:           ║
║ - Playwright: 1.22.0                                                                       ║
║ -     Docker: 1.21.0                                                                       ║
║                                                                                            ║
║ Either:                                                                                    ║
║ - (recommended) use docker image "mcr.microsoft.com/playwright:v1.22.0-focal"              ║
║ - (alternative 1) run the following command inside docker to install missing dependencies: ║
║                                                                                            ║
║     npx playwright install-deps                                                            ║
║                                                                                            ║
║ - (alternative 2) use Aptitude inside docker:                                              ║
║                                                                                            ║
║     apt-get install libgbm1                                                                ║
║                                                                                            ║
║ <3 Playwright Team                                                                         ║
╚════════════════════════════════════════════════════════════════════════════════════════════╝
    at file:///root/a.mjs:3:10 {
  name: 'Error'
}```

Fixes #12796

dgozman commented on packages/playwright-core/src/utils/dependencies.ts:
```diff
@@ -183,11 +201,36 @@ export async function validateDependenciesLinux(sdkLanguage: string, linuxLddDir
-    // Happy path: known dependencies are missing for browsers.
+  // Ignore patch versions when comparing docker container version and Playwright version:
+  // we **NEVER** roll browsers in patch releases, so native dependencies do not change.
+  if (dockerInfo && !dockerInfo.driverVersion.startsWith(utils.getPlaywrightVersion(true /* majorMinorOnly */))) {
```
```suggestion
  if (dockerInfo && !dockerInfo.driverVersion.startsWith(utils.getPlaywrightVersion(true /* majorMinorOnly */) + '.')) {
```

To avoid 1.200 vs 1.20 mismatch 😄

aslushnikov committed:
Update packages/playwright-core/src/utils/dependencies.ts

Co-authored-by: Dmitry Gozman <dgozman@gmail.com>
